src/garage/_environment.py

```diff
@@ -1,6 +1,222 @@
 """Base Garage Environment API."""
 
 import abc
+import collections
+import math
+
+import akro
+import numpy as np
+
+from garage import InOutSpec, StepType
+
+
+class EnvSpec(InOutSpec):
+    """Describes the action and observation spaces of an environment.
+
+    Args:
+        observation_space (akro.Space): The observation space of the env.
+        action_space (akro.Space): The action space of the env.
+        max_episode_length (int): The maximum number of steps allowed in an
+            episode.
+
+    """
+
+    def __init__(self,
+                 observation_space,
+                 action_space,
+                 max_episode_length=math.inf):
+        self._max_episode_length = max_episode_length
+        super().__init__(action_space, observation_space)
+
+    @property
+    def action_space(self):
+        """Get action space.
+
+        Returns:
+            akro.Space: Action space of the env.
+
+        """
+        return self.input_space
+
+    @property
+    def observation_space(self):
+        """Get observation space of the env.
+
+        Returns:
+            akro.Space: Observation space.
+
+        """
+        return self.output_space
+
+    @action_space.setter
+    def action_space(self, action_space):
+        """Set action space of the env.
+
+        Args:
+            action_space (akro.Space): Action space.
+
+        """
+        self._input_space = action_space
+
+    @observation_space.setter
+    def observation_space(self, observation_space):
+        """Set observation space of the env.
+
+        Args:
+            observation_space (akro.Space): Observation space.
+
+        """
+        self._output_space = observation_space
+
+    @property
+    def max_episode_length(self):
+        """Get max episode steps.
+
+        Returns:
+            int: The maximum number of steps that an episode
+
+        """
+        return self._max_episode_length
+
+
+class EnvStep(
```

Comment:
all of this validation can get expensive on a per-step level, and will be repeated in TrajectoryBatch or TimeStepBatch. I think we should leave this as a simple namedtuple